core/src/apps/cardano/address.py

```diff
@@ -1,137 +1,249 @@
+        address = _derive_base_address(
+            keychain,
+            address_header,
+            parameters.path,
+            parameters.staking_key_path,
+            parameters.staking_key_hash,
+        )
+    elif parameters.address_type == CardanoAddressType.ENTERPRISE_ADDRESS:
+        address = _derive_enterprise_address(keychain, address_header, parameters.path)
+    elif parameters.address_type == CardanoAddressType.POINTER_ADDRESS:
+        address = _derive_pointer_address(
+            keychain, address_header, parameters.path, parameters.certificate_pointer,
+        )
+    elif parameters.address_type == CardanoAddressType.REWARD_ADDRESS:
+        address = _derive_reward_address(keychain, address_header, parameters.path)
+    else:
+        raise ValueError("Invalid address type '%s'!" % parameters.address_type)
 
-    if not isinstance(address_protocol_magic, int):
-        return False
+    return address
 
-    return address_protocol_magic == protocol_magic
 
+def _is_shelley_address_path(path: List[int]) -> bool:
+    return path[0] == SHELLEY_PURPOSE
 
-def validate_full_path(path: list) -> bool:
-    """
-    Validates derivation path to fit 44'/1815'/a'/{0,1}/i,
-    where `a` is an account number and i an address index.
-    The max value for `a` is 20, 1 000 000 for `i`.
-    The derivation scheme v1 allowed a'/0/i only,
-    but in v2 it can be a'/1/i as well.
-    """
-    if len(path) != 5:
-        return False
-    if path[0] != 44 | HARDENED:
-        return False
-    if path[1] != 1815 | HARDENED:
-        return False
-    if path[2] < HARDENED or path[2] > 20 | HARDENED:
-        return False
-    if path[3] != 0 and path[3] != 1:
-        return False
-    if path[4] > 1000000:
-        return False
-    return True
 
+def _create_address_header(address_type: int, network_id: int) -> bytes:
+    if address_type == CardanoAddressType.BYRON_ADDRESS:
+        """
+        Byron addresses don't have an explicit header in the Shelley
+        spec. However, thanks to their CBOR structure they always start with
+        0b1000 - the byron address id. This is no coincidence.
+        The Shelley address headers are purposefully built around these
+        starting bits of the byron address.
+        """
+        raise ValueError("Byron address does not contain an explicit header!")
+
+    address_id = _get_address_id(address_type)
+    header = address_id << 4 | network_id
 
-def _address_hash(data) -> bytes:
-    data = cbor.encode(data)
-    data = hashlib.sha3_256(data).digest()
-    res = hashlib.blake2b(data=data, outlen=28).digest()
-    return res
+    return bytes([header])
 
 
-def _get_address_root(node, payload):
-    extpubkey = remove_ed25519_prefix(node.public_key()) + node.chain_code()
-    if payload:
-        payload = {1: cbor.encode(payload)}
+def _get_address_id(address_type: int) -> int:
+    if address_type == CardanoAddressType.BYRON_ADDRESS:
```

Comment:
why don't we create this mapping as dict?

Reply:
I've unified it. CardanoAddressType contains the header values.